aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -587,5 +587,91 @@ Tensor &nuclear_norm_out(Tensor& result, const Tensor& self, bool keepdim) {
   return at::sum_out(result, std::get<1>(at::svd(self)), 0, keepdim);
 }
 
+static inline Tensor _chain_matmul_general(TensorList matrices, std::vector<std::vector<int64_t>>& order, int64_t i, int64_t j) {
+  if (i == j)
+    return matrices[i];
+  else
+    return at::mm(_chain_matmul_general(matrices, order, i, order[i][j]), _chain_matmul_general(matrices, order, order[i][j] + 1, j));
+}
+
+// Why the separate implementation for 3 matrices?
+// The logic for three matrices is much faster when done directly
+// Requires 1 comparison to 4 comparisons and lesser arithmetic operations
+static inline Tensor _chain_matmul_three_matrices(TensorList matrices) {
+  int64_t a = matrices[0].size(0);  // This is the first dimension
+  int64_t b = matrices[1].size(0);  // This is the common dimension between the first two matrices
+  int64_t c = matrices[2].size(0);  // This is the common dimension between the last two matrices
+  int64_t d = matrices[2].size(1);  // This is the last dimension
+
+  // The matrices are of size (a x b), (b x c), (c x d)
+  // cost_1 is the cost of parenthesizing (a x b) and (b x c) and then combining (c x d)
+  // cost_2 is the cost of parenthesizing (b x c) and (c x d) and then combining (a x b)
+  int64_t cost_1 = (a * c) * (b + d);
+  int64_t cost_2 = (b * d) * (a + c);
+
+  if (cost_1 > cost_2) {
+    return at::mm(matrices[0], at::mm(matrices[1], matrices[2]));
+  } else {
+    return at::mm(at::mm(matrices[0], matrices[1]), matrices[2]);
+  }
+}
+
+Tensor chain_matmul(TensorList matrices) {
+  checkAllSameDim(matrices, 2);
+
+  if (matrices.size() == 1) {
+    return matrices[0];
+  } else if (matrices.size() == 2) {
+    return at::mm(matrices[0], matrices[1]);
+  } else if (matrices.size() == 3) {
+    return _chain_matmul_three_matrices(matrices);
+  } else {
+
+    // Following the algorithm in Chapter 15.2 : Introduction to Algorithms, Cormen et al.
+    // Minor modifications have been made to accommodate zero-indexing
+    auto n = matrices.size();
+
+    // Dim vector - the length of which is n + 1. Note that for matrix multiplication, there
+    // needs to a common dimension between the multiplicands, hence for n matrices, there are
+    // n + 1 values. The values p_{i} and p_{i + 1} correspond to the dimensions of matrix i in
+    // the chain (zero-indexed)
+    std::vector<int64_t> p;
+    p.push_back(matrices[0].size(0));
+    for (int64_t i = 0; i < n; i++) {
+      p.push_back(matrices[i].size(1));
+    }
+
+    // Cost matrix - an element m[i, j] of this matrix corresponds to the minimum cost of
+    // parenthesizing matrices A_{i} to A_{j}. By this definition m[i, i] = 0 for all i
+    // m[i, j] is filled using the substructure property of the algorithm, meaning:
+    // m[i, j] = min_{i <= k < j} m[i, k] + m[k, j] + p_{i-1}p_{k}p_{j}
+    std::vector<std::vector<int64_t>> m(n, std::vector<int64_t>(n, 0));
```

Comment:
actually, could you use `std::numeric_limits<int64_t>::max()` here instead of `0`?

Reply:
I have done it. I had to reinitialize the diagonals using the outer loop though, but I think that’s cheaper.

Reply:
The change caused overflow, I reverted it.